python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py

```diff
@@ -60,6 +60,7 @@ def asset(
     key_prefix: Optional[CoercibleToAssetKeyPrefix] = None,
     ins: Optional[Mapping[str, AssetIn]] = ...,
     non_argument_deps: Optional[Union[Set[AssetKey], Set[str]]] = ...,
+    upstream_assets: Optional[Set[CoercibleToAssetKey]] = ...,
```

Comment:
Discussion: What should the type signature of the new parameter be?
Some options:
* `Optional[Set[CoercibleToAssetKey]]`
* `Optional[Set[Union[AssetKey, str]]]`
* `Optional[Set[Union[CoercibleToAssetKey, UpstreamAsset]]]` Where `UpstreamAsset` (name tbd) is a new class allowing you to specify asset key, partition mapping, etc
* `Optional[Set[Union[AssetKey, str, UpstreamAsset]]]`

related question - `Set` vs `Sequence` with some runtime checking for duplicated

Reply:
We've discussed a `Set` or `Sequence` containing some combination of the following:
* `CoercibleToAssetKey`
* `str`
* `List[str]`
* `AssetsDefinition`
* `UpstreamAsset` (name tbd) - a new class allowing you to specify asset key, partition mapping, etc

Reply:
I am going to write up a PR that explores passing `AssetDefinitions` around

Reply:
Also wrote up a rough PR exploring the `UpstreamAsset` class https://github.com/dagster-io/dagster/pull/14908

Reply:
Some reflections after writing those two PRs:
* passing asset definitions is very ergonomic. When I wrote the `UpstreamAsset` example, i missed using the definition python symbol
* If we think that passing around the full definition object is a bad pattern, we could instead have users pass `my_asset.key` which still lets you work with symbols 
* maybe a bit out there, but what are the consequences of allowing `AssetsDefinition` as a `CoercibleToAssetKey`? We'd need to decide how to handle multi_assets (either make a dependency on all assets w/in the multi asset or explicitly disallow it)

Reply:
If we allow `AssetsDefinition` (which definitely has a certain appeal) I think the usage of `Set` versus `Sequence` becomes increasingly problematic. I think it is more reflective in reality to say we take a sequence of things and then do runtime checks/processing to unique-ify. 

The previous `non_argument_deps` as typed the way it was for good reason, since strings and asset keys don't play together nicely in a set:

```
from dagster import AssetKey

string = "foo"
key = AssetKey("foo")
set([string, key])
# prints {AssetKey(['foo']), 'foo'} in interpreter
```

So if we want to mix and match these things it needs to be a `Sequence` IMO

Reply:
The other behavior to consider if `AssetsDefinition` is provided is to only add dependencies to the "leaf" assets (those with no downstream dependences within that AssetsDefinition) to avoid cluttering the dependency graph.